sdk/cxx/include/sawtooth_sdk.h

```diff
@@ -0,0 +1,258 @@
+/* Copyright 2017 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <map>
+#include <memory>
+#include <list>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace sawtooth {
+
+// Utility types to allow for managed instances of std::string on the heap.
+typedef std::unique_ptr<std::string> StringUPtr;
+typedef std::shared_ptr<std::string> StringPtr;
+
+
+enum TransactionHeaderField
+{
+    TransactionHeaderBatcherPublicKey = 1,
+    TransactionHeaderStringDependencies,
+    TransactionHeaderFamilyName,
+    TransactionHeaderFamilyVersion,
+    TransactionHeaderInputs,
+    TransactionHeaderNonce,
+    TransactionHeaderOutputs,
+    TransactionHeaderPayloadSha512,
+    TransactionHeaderSignerPublicKey
+};
+
+class TransactionHeaderIF
+{
+public:
+    virtual ~TransactionHeaderIF(){};
+
+    virtual int GetCount(TransactionHeaderField field) = 0;
+    virtual const ::std::string& GetValue(TransactionHeaderField field, int index = 0) = 0;
+};
+
+typedef std::unique_ptr<TransactionHeaderIF> TransactionHeaderUPtr;
+typedef std::shared_ptr<TransactionHeaderIF> TransactionHeaderPtr;
+
+// The transaction data for a Transaction Processing request.
+class Transaction final {
+ public:
+    Transaction(TransactionHeaderPtr header, StringPtr payload, StringPtr signature):
+            header_(header), payload_(payload), signature_(signature) {
+    }
+
+    Transaction (const Transaction&) = delete;
+    Transaction (const Transaction&&) = delete;
+    Transaction& operator= (const Transaction&) = delete;
+
+    const TransactionHeaderPtr header() const {
+        return this->header_;
+    }
+
+    const std::string& payload() const {
+        return *(this->payload_);
+    }
+
+    const std::string& signature() const {
+        return *(this->signature_);
+    }
+
+ private:
+    TransactionHeaderPtr header_;
+    StringPtr payload_;
+    StringPtr signature_;
+};
+typedef std::unique_ptr<Transaction> TransactionUPtr;
+
+
+// Provides access to the global state
+class GlobalStateIF {
```

Comment:
There should be only one implementation of state, which suggests it should not be an interface.

Reply:
The primary goal of this interface is to isolate app developers from underlying details of GlobalState implementation, no SDK code is required for building a TP.